src/v/kafka/data/cloud_topic_partition.cc

```diff
@@ -0,0 +1,599 @@
+ * by the Apache License, Version 2.0
+ */
+#include "kafka/data/cloud_topic_partition.h"
+
+#include "cloud_storage/types.h"
+#include "cloud_topics/app.h"
+#include "cloud_topics/data_plane_api.h"
+#include "cloud_topics/dl_placeholder.h"
+#include "cloud_topics/extent_meta.h"
+#include "cluster/partition.h"
+#include "cluster/rm_stm.h"
+#include "cluster/types.h"
+#include "kafka/protocol/batch_reader.h"
+#include "kafka/protocol/errors.h"
+#include "logger.h"
+#include "model/fundamental.h"
+#include "model/record.h"
+#include "model/record_batch_reader.h"
+#include "model/record_batch_types.h"
+#include "model/timeout_clock.h"
+#include "raft/consensus_utils.h"
+#include "raft/errc.h"
+#include "raft/replicate.h"
+#include "storage/log_reader.h"
+#include "storage/record_batch_builder.h"
+#include "storage/types.h"
+
+#include <seastar/core/circular_buffer.hh>
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/shared_ptr.hh>
+#include <seastar/coroutine/as_future.hh>
+
+#include <chrono>
+#include <iterator>
+#include <optional>
+#include <system_error>
+
+namespace kafka {
+
+namespace {
+
+struct placeholder_batches_with_size {
+    ss::circular_buffer<model::record_batch> batches;
+    // Total size of all referenced data
+    size_t extent_size{0};
+};
+
+static model::record_batch make_placeholder_batch(
+  const model::record_batch_header& hdr,
+  const experimental::cloud_topics::extent_meta& extent) {
+    // The caller is supposed to use the correct batch header
+    // that matches the extent.
+    vassert(
+      hdr.base_offset == kafka::offset_cast(extent.base_offset),
+      "Base offset of the extent {} doesn't match the data batch {}",
+      extent.base_offset,
+      hdr.base_offset);
+    vassert(
+      hdr.last_offset() == kafka::offset_cast(extent.last_offset),
+      "Base offset of the extent {} doesn't match the data batch {}",
+      extent.base_offset,
+      hdr.base_offset);
+
+    experimental::cloud_topics::dl_placeholder placeholder{
+      .id = extent.id,
+      .offset = extent.first_byte_offset,
+      .size_bytes = extent.byte_range_size,
+    };
+
+    storage::record_batch_builder builder(
+      model::record_batch_type::dl_placeholder, hdr.base_offset);
+
+    auto first_key = serde::to_iobuf(
+      experimental::cloud_topics::dl_placeholder_record_key::payload);
+
+    auto first_value = serde::to_iobuf(placeholder);
+
+    // In case of a placeholder batch the first record contains the
+    // actual placeholder and the remaining records are empty. The remaining
+    // records are added to avoid confusing any other code that may expect
+    // that the number of records in the batch is equal to the number of
+    // offsets in the header.
+    builder.add_raw_kv(std::move(first_key), std::move(first_value));
+
+    for (int i = 1; i < hdr.record_count; ++i) {
+        builder.add_raw_kv(std::nullopt, std::nullopt);
+    }
```

Comment:
Do we need to preserve timestamps from the original batch?

Reply:
> i updated this code path to propagate some additional fields (timestamps, producer id/epoch)

And why wouldn't we retain the entire header? It's practically more work to cherry pick?